packages/compass-connections/src/connections-manager.ts

```diff
@@ -0,0 +1,223 @@
+  ) => void;
+};
+
+export enum ConnectionStatus {
+  Disconnected = 'disconnected',
+  Connecting = 'connecting',
+  Connected = 'connected',
+  Failed = 'failed',
+}
+
+export class ConnectionsManager extends EventEmitter {
+  private connectionAttempts = new Map<ConnectionInfoId, ConnectionAttempt>();
+  private connectionStatuses = new Map<ConnectionInfoId, ConnectionStatus>();
+  private dataServices = new Map<ConnectionInfoId, DataService>();
+
+  constructor(
+    private readonly __TEST_CONNECT_FN?: ConnectFn,
+    private readonly logger?: LoggerAndTelemetry['log']['unbound']
+  ) {
+    super();
+  }
+
+  getDataServiceForConnection(
+    connectionInfoId: ConnectionInfoId
+  ): DataService | undefined {
+    return this.dataServices.get(connectionInfoId);
+  }
+
+  statusOf(connectionInfoId: ConnectionInfoId): ConnectionStatus {
+    return (
+      this.connectionStatuses.get(connectionInfoId) ??
+      ConnectionStatus.Disconnected
+    );
+  }
+
+  cancelAllConnectionAttempts(): void {
+    for (const connectionInfoId of this.connectionAttempts.keys()) {
+      this.cancelConnectionAttempt(connectionInfoId);
+    }
+  }
+
+  cancelConnectionAttempt(connectionInfoId: ConnectionInfoId): void {
+    const connectionAttempt = this.connectionAttempts.get(connectionInfoId);
+    connectionAttempt?.cancelConnectionAttempt();
+    this.connectionAttempts.delete(connectionInfoId);
+    this.notifyConnectionAttemptCancelled(connectionInfoId);
+  }
+
+  /**
+   * @param connectionInfo The adjusted ConnectionInfo object that already has
+   * parameters such as appName.
+   */
+  async connect(connectionInfo: ConnectionInfo): Promise<void> {
+    try {
+      if (
+        [ConnectionStatus.Connecting, ConnectionStatus.Connected].includes(
+          this.statusOf(connectionInfo.id)
+        )
+      ) {
+        return;
+      }
+
+      this.notifyConnectionAttemptStarted(connectionInfo.id);
+      const connectionAttempt = createConnectionAttempt({
+        logger: this.logger,
+        connectFn: this.__TEST_CONNECT_FN,
+      });
+      this.connectionAttempts.set(connectionInfo.id, connectionAttempt);
+
+      const dataService = await connectionAttempt.connect(
+        connectionInfo.connectionOptions
+      );
+
+      if (!dataService || connectionAttempt.isClosed()) {
+        return;
+      }
+
+      this.connectionAttempts.delete(connectionInfo.id);
+      this.dataServices.set(connectionInfo.id, dataService);
+      this.notifyConnectionAttemptSuccessful(connectionInfo.id, dataService);
+    } catch (error) {
+      this.notifyConnectionAttemptFailed(connectionInfo.id, error);
+      throw error;
+    }
+  }
+
+  async closeConnection(connectionInfoId: ConnectionInfoId): Promise<void> {
+    const dataService = this.dataServices.get(connectionInfoId);
+    if (dataService) {
+      await dataService.disconnect();
+      this.dataServices.delete(connectionInfoId);
+      this.notifyConnectionDisconnected(connectionInfoId);
+    }
```

Comment:
This seems a bit loose of an interface, if `closeConnection` is called while connection is still in progress, it will be a noop and you will continue connecting. If we want two separate interfaces for cancelling an in-progress connection and disconnecting a connection that was established, maybe this should be a bit more strict, more specifically this should probably check the current status and maybe throw in some cases where the method is being misused. Alternatively this and `cancelConnectionAttempt` can be one method that handles both cases, I don't think we have any particular need for this separation, but I might be missing something